File: mid/Tree/230. Kth Smallest Element in a BST/230. Kth Smallest Element in a BST (Stack+Iterate).cpp
// https://bclin.tw/2022/07/06/leetcode-230/
// Time Complexity: O(N)
// Space Complexity: O(N)
// Stack + Iterate

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode*> node_stack;
        int nodeCount = 0;

        while(true) {
            while(root) {
                node_stack.push(root);
                root = root->left;
            }
            root = node_stack.top();
            nodeCount += 1;
            node_stack.pop();
            if(nodeCount == k){
                break;
            }
            root = root->right;
        }
        return root->val;
    }
};

File: mid/Tree/230. Kth Smallest Element in a BST/230. Kth Smallest Element in a BST (Stack+Iterate).py
# 
# Time Complexity: O(N)
# Space Complexity: O(H)
# Stack + Iterate
# Beats 100%

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        stack = []
        nodeCount = 0
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            nodeCount += 1 
            if nodeCount == k:
                break
            root = root.right
        return root.val